tools/queries.cpp

```diff
@@ -44,85 +44,156 @@
+            aggregated_query_times.push_back(mean);
+        }
+    } else if (aggregation_type == AggregationType::Median) {
+        for (auto const& query_times: times_per_query) {
+            auto sorted_query_times = query_times;
+            std::sort(sorted_query_times.begin(), sorted_query_times.end());
+            std::size_t sample_count = sorted_query_times.size();
+            double median = 0;
+            if (sample_count % 2 == 1) {
+                median = sorted_query_times[sample_count / 2];
+            } else {
+                median =
+                    (sorted_query_times[sample_count / 2] + sorted_query_times[sample_count / 2 - 1])
+                    / 2;
+            }
+            aggregated_query_times.push_back(median);
+        }
+    } else if (aggregation_type == AggregationType::Max) {
+        for (auto const& query_times: times_per_query) {
+            aggregated_query_times.push_back(*std::max_element(query_times.begin(), query_times.end()));
+        }
     }
+    std::sort(aggregated_query_times.begin(), aggregated_query_times.end());
+    return aggregated_query_times;
+}
+
+void print_stats(AggregationType aggregation_type, std::vector<std::size_t> const& query_times) {
+    double mean =
+        std::accumulate(query_times.begin(), query_times.end(), double()) / query_times.size();
+    double q50 = query_times[query_times.size() / 2];
+    double q90 = query_times[90 * query_times.size() / 100];
+    double q95 = query_times[95 * query_times.size() / 100];
+    double q99 = query_times[99 * query_times.size() / 100];
+    stats_line(std::cerr)("aggregated_by", to_string(aggregation_type))("mean", mean)("q50", q50)("q90", q90)("q95", q95)("q99", q99);
 }
 
-template <typename Functor>
-void op_perftest(
-    Functor query_func,
+template <typename Fn>
+void extract_times(
+    Fn query_func,
     std::vector<Query> const& queries,
     std::vector<Score> const& thresholds,
     std::string const& index_type,
     std::string const& query_type,
     size_t runs,
     std::uint64_t k,
-    bool safe
+    bool safe,
+    AggregationType aggregate_by,
+    bool summary_only,
+    std::ostream& os
 ) {
-    std::vector<double> query_times;
-    std::size_t num_reruns = 0;
+    std::vector<std::vector<std::size_t>> times_per_query(
+        queries.size(), std::vector<std::size_t>(runs)
+    );
+    std::size_t corrective_rerun_count = 0;
     spdlog::info("Safe: {}", safe);
 
+    // Note: each query is measured once per run, so the set of queries is
+    // measured independently in each run.
     for (size_t run = 0; run <= runs; ++run) {
-        size_t idx = 0;
+        size_t query_idx = 0;
         for (auto const& query: queries) {
             auto usecs = run_with_timer<std::chrono::microseconds>([&]() {
-                uint64_t result = query_func(query, thresholds[idx]);
+                uint64_t result = query_func(query, thresholds[query_idx]);
                 if (safe && result < k) {
```

Comment:
@elshize, do you think it is convenient to define another path to avoid performing this check for each query run? I'm referring to the cost of evaluating it (which, in any case, is a "constant cost" that will be added to all runs).